gfx/3d: improve performance

rr- opened the pull request:
#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TR1X/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

Resolves #2325. (Hopefully)

rr- committed:
gfx/3d: preallocate vertices

Resolves #2325.

aredfan commented:
Thank you for trying to fix the issue.  Sorry to be the bearer of bad news but the PR build doesn't fix the issue.

rr- commented:
In this case I think the only option is to do general performance optimizations, which I can implement after 0.9.

rr- committed:
gfx/3d: reduce amount of flushes

rr- commented:
@aredfan @lahm86 this is ready for a re-review.

aredfan reviewed (approved):
> @aredfan @lahm86 this is ready for a re-review.

The Floating Islands level is loading smoothly again, and TR1X looks good as well.

Thank you for your perseverance. 👍

walkawayy commented on docs/tr2/CHANGELOG.md:
```diff
@@ -33,7 +33,7 @@
 - fixed Skidoo snow wake effects at slow speeds (#2324, regression from 0.6)
 - fixed software renderer skybox occlusion issues (#2343, regression from 0.7)
-- fixed Floating Islands losing some frames if played first (#2325, regression from 0.7)
+- improved slightly rendering performance in big rooms (#2325)
```
Idk if you want to match TR1X's changelog or not, but this wording is a little awkward. Normally it's `slightly improved` but you want the verb first in the changelog. I think `improved rendering performance slightly in big rooms` would work. Sorry for the nitpick. Great fix!

rr- replied:
Oh, thank you for spotting this! Indeed, I had a better wording for this, but it got lost in the rebases. I'll try to recover it, since I need to do a small fix anyway.

rr- replied:
Fixed

rr- commented:
@lahm86 fixed the mesh 0 issue; @aredfan LMK if the fix broke anything.

aredfan reviewed (approved):
Still looks good. 👍

aredfan commented:
Actually I just noticed this issue, see the versioning at the bottom right during fade-in.
![Screenshot](https://github.com/user-attachments/assets/3d3263db-7cb0-4050-9634-7ab7e4b6a893)

End of level stats after Opera House and CUT2:
![20250121_203844_Opera_House](https://github.com/user-attachments/assets/7aeba679-006a-435d-97bf-e698de836a32)

Credits and final stats:
![20250121_204002_Home_Sweet_Home](https://github.com/user-attachments/assets/5d99ca00-75c7-4127-913b-4944469f04a6)

If you open/close the inventory a few times - the stopwatch starts showing odd textures.
![20250121_205009_Laras_Home](https://github.com/user-attachments/assets/c315c011-646a-43d3-9e50-74d2ea0d7a1c)

rr- commented:
@aredfan I changed the approach to match a similar optimization we had in TR1. Measuring the impact on TR1 shows no improvement there, so I've removed the changelog entry. LMK  if it's working as intended. Thanks

aredfan commented:
@rr- All the prior issues are now fixed. I also give TR1X another check as well as TR2X, and both builds are working as expected.

LGTM 👍
